nixos/modules/services/x11/extra-layouts.nix

```diff
@@ -0,0 +1,150 @@
+  ''
+      # install layout file
+      cp ${layout.filepath} symbols/${name}
+
+      # patch makefiles
+      ${pkgs.ed}/bin/ed -v symbols/Makefile.am <<EOF
+      /symbols_DATA =
+      a
+      ${name} \\
+      .
+      w
+      EOF
+      ${pkgs.ed}/bin/ed -v symbols/Makefile.in <<EOF
+      /symbols_DATA =
+      a
+      ${name} \\
+      .
+      w
+      EOF
+
+      # add layout description
+      ${pkgs.ed}/bin/ed -v rules/base.xml.in <<EOF
+      /<\/layoutList>
+      -
+      a
+      <layout>
+        <configItem>
+          <name>${name}</name>
+          <_shortDescription>${name}</_shortDescription>
+          <_description>${layout.description}</_description>
+          <languageList>
+            ${lib.concatMapStrings (lang: "<iso639Id>${lang}</iso639Id>\n") layout.languages}
+          </languageList>
+        </configItem>
+        <variantList/>
+      </layout>
+      .
+      w
+      EOF
+      cp rules/base.xml.in rules/base.evdev.xml.in
+  '';
+
+  xkb_patched = pkgs.xorg.xkeyboardconfig.overrideAttrs (old: {
+    buildInputs = old.buildInputs ++ [ pkgs.automake ];
+    postPatch = with lib; concatStrings (mapAttrsToList patchIn layouts);
+  });
+
+
+in
+
+{
+
+  ###### interface
+
+  options.services.xserver = {
+    extraLayouts = mkOption {
+      type = types.attrsOf (types.submodule layoutOpts);
+      default = {};
+      example = literalExample
+      ''
+        {
+          mine = {
+            description = "My custom xkb layout.";
+            languages = [ "eng" ];
+            filepath = /path/to/my/layout;
+          };
+        }
+      '';
+      description = ''
+        Extra custom layouts that will be included in the xkb configuration.
+        Information on how to create a new layout can be found here:
+        https://www.x.org/releases/current/doc/xorg-docs/input/XKB-Enhancing.html#Defining_New_Layouts
+      '';
+    };
+
+  };
+
+  ###### implementation
+
+  config = mkIf (layouts != { }) {
+
+    nixpkgs.config.packageOverrides = super: {
+      xorg = super.xorg // rec {
+        xorgserver = super.xorg.xorgserver.overrideAttrs (old: {
+          configureFlags = old.configureFlags ++ [
+            "--with-xkb-bin-directory=${xkbcomp}/bin"
+            "--with-xkb-path=${xkb_patched}/share/X11/xkb"
+          ];
+        });
+
+        setxkbmap = super.xorg.setxkbmap.overrideAttrs (old: {
+          postInstall =
+            ''
+              mkdir -p $out/share
+              ln -sfn ${xkb_patched}/etc/X11 $out/share/X11
+            '';
+        });
+
+        xkbcomp = super.xorg.xkbcomp.overrideAttrs (old: {
+          configureFlags = "--with-xkb-config-root=${xkb_patched}/share/X11/xkb";
+        });
+
+      };
+
+      ckbcomp = super.ckbcomp.override {
+        xkeyboard_config = xkb_patched;
+      };
+
+      xkbvalidate = super.xkbvalidate.override {
+        libxkbcommon = super.libxkbcommon.override {
+          xkeyboard_config = xkb_patched;
```

Comment:
Having to override everything that uses xkb manually isn't very nice. By changing xorg's `xkeyboard_config` attribute you should be able to change those all with one setting (unless I'm missing something).

Reply:
In theory yes, but that would require to recompile a huge amount of software. This trick minimizes that to only a few packages.

Reply:
Ah darn :/